### PR TITLE
Refactor read(data|csv|yaml|envfile|json) to use new DATAFILETYPE enums instead of parsing strings

### PR DESCRIPTION
**Pull request by karlhto**
This was originally just meant for `validdata`, but I wanted to make the existing functions a bit more modular (and hopefully more readable as well).

### CODE REVIEW

**Comment by karlhto** [comment]
@cf-bottom Jenkins with exotics please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4314/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4314/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4314/

**Comment by karlhto** [comment]
@cf-bottom Jenkins with exotics

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4315/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4315/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4315/

**Comment by karlhto** [comment]
@cf-bottom Jenkins please

**Comment by karlhto** [comment]
Only change is that `libntech` is updated to latest master. : )

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4343/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4343/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4343/

**Comment by karlhto** [comment]
@cf-bottom Jenkins

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4345/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4345/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4345/

File: libpromises/evalfunction.c [hunk 1]
```diff
@@ -6634,68 +6633,106 @@ static FnCallResult FnCallReadRealList(EvalContext *ctx, ARG_UNUSED const Policy
...
 
-    // JsonReadDataFile reads key-value pairs from a json, yaml, csv or
-    // env file and puts them in a JSON element
-    // requested_mode is used to specify the input file type
-    // requested_mode is set based on file extension or function name
-    // example: readyaml -> "YAML" , ".csv" -> "CSV" etc.
-    // defaults to "JSON"
-    const char *requested_mode = NULL;
-    if (strcmp(fp->name, "readdata") == 0)
+    // readdata gets rid of the size
```
Comment (vpodzime) [inline libpromises/evalfunction.c]: This `readdata` reference is confusing here.

File: libpromises/evalfunction.c [hunk 2]
```diff
@@ -6634,68 +6633,106 @@ static FnCallResult FnCallReadRealList(EvalContext *ctx, ARG_UNUSED const Policy
...
+    const size_t size_max = IntFromString("inf");
+    const char *const mode_string = RlistScalarValue(args->next);
+    DataFileType requested_mode = DATAFILETYPE_UNKNOWN;
+    if (strcmp("auto", mode_string) == 0)
```
Comment (vpodzime) [inline libpromises/evalfunction.c]: Please use `SafeStringEqual` wherever possible.

File: libpromises/evalfunction.c [hunk 3]
```diff
@@ -6634,68 +6633,106 @@ static FnCallResult FnCallReadRealList(EvalContext *ctx, ARG_UNUSED const Policy
...
-    // JsonReadDataFile reads key-value pairs from a json, yaml, csv or
-    // env file and puts them in a JSON element
-    // requested_mode is used to specify the input file type
-    // requested_mode is set based on file extension or function name
-    // example: readyaml -> "YAML" , ".csv" -> "CSV" etc.
-    // defaults to "JSON"
-    const char *requested_mode = NULL;
-    if (strcmp(fp->name, "readdata") == 0)
+    // readdata gets rid of the size
+    const size_t size_max = IntFromString("inf");
```
Comment (vpodzime) [inline libpromises/evalfunction.c]: Isn't there a numeric constant for this?

**Comment by karlhto** [comment]
@cf-bottom Jenkins

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4361/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4361/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4361/

**Comment by karlhto** [comment]
Last push was just to fix correct commit email. No additional changes.

File: libpromises/evalfunction.c [hunk 4]
```diff
@@ -6634,68 +6633,103 @@ static FnCallResult FnCallReadRealList(EvalContext *ctx, ARG_UNUSED const Policy
     return ReadList(ctx, fp, args, CF_DATA_TYPE_REAL);
 }
 
+static FnCallResult _ReadDataGeneric(const char *const fname,
```
Comment (olehermanse) [inline libpromises/evalfunction.c]: I think we've mentioned this before, but: We don't use `_` for internal / static functions.

**Comment by olehermanse** [comment]
Looks green

**Comment by olehermanse** [comment]
Cherry-pick to 3.15.x: https://github.com/cfengine/core/pull/3945

**Pull request merged**